File: 2 - Sorting/C/merge_sort.c
/* C program for Merge Sort */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Merges two subarrays of arr[].
// First subarray is arr[l..m]
// Second subarray is arr[m+1..r]
void merge(int arr[], int l, int m, int r) {
	int i, j, k;
	int n1 = m - l + 1;
	int n2 = r - m;

	int L[n1], R[n2]; /* create temp arrays */

	/* Copy data to temp arrays L[] and R[] */
	for (i = 0; i < n1; i++)
		L[i] = arr[l + i];
	for (j = 0; j < n2; j++)
		R[j] = arr[m + 1 + j];

	/* Merge the temp arrays back into arr[l..r]*/
	i = 0; // Initial index of first subarray
	j = 0; // Initial index of second subarray
	k = l; // Initial index of merged subarray
	while (i < n1 && j < n2) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		}
		else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	/* Copy the remaining elements of L[], if there	are any */
	while (i < n1) {
		arr[k] = L[i];
		i++;
		k++;
	}

	/* Copy the remaining elements of R[], if there are any */
	while (j < n2) {
		arr[k] = R[j];
		j++;
		k++;
	}
}

/* l is for left index and r is right index of the
sub-array of arr to be sorted */
void mergeSort(int arr[], int l, int r)
{
	if (l < r) {
		// Same as (l+r)/2, but avoids overflow for
		// large l and h
		int m = l + (r - l) / 2;

		// Sort first and second halves
		mergeSort(arr, l, m);
		mergeSort(arr, m + 1, r);

		merge(arr, l, m, r);
	}
}

/* UTILITY FUNCTIONS */
/* Function to print an array */
void printArray(int A[], int size)
{
	int i;
	for (i = 0; i < size; i++)
		printf("%d ", A[i]);
	printf("\n");
}


int *read_input(char FILENAME[], int *ARRAY_LENGTH){
	FILE *file;
    static int array[30][30];
    char row_data[300];

    // MENARUH DATA INPUT KE DALAM ARRAY
    file = fopen(FILENAME, "r");
    fgets(row_data, 200, file);
    char *TOKEN = strtok(row_data, ",");
    int i = 0;
    while (TOKEN != NULL) {
        array[1][i] = atoi(TOKEN);
        // printf("%d ", array[1][i]);
        TOKEN = strtok(NULL, ",");
        i++;
    }
	// SETTING LENGTH DARI ARRAY / ARRAY SIZE
	*ARRAY_LENGTH = i;
    fclose(file);
    // printf("\n");

	return array[1];
}


void write_to_txt(int arr[], int arr_len){
	FILE *newfile;
	char namafile[54];
	char word[20];
	
	printf("\nMasukkan nama file yang ingin disave: ");
    scanf("%s", namafile);
	strcat(namafile, ".txt");
	newfile = fopen(namafile, "w");
	for (int i = 0; i < arr_len; i++){
		sprintf(word, "%d", arr[i]);
		strcat(word, ",");
		fprintf(newfile, word);
	}
	fclose(newfile);
}


/* Driver code */
void main()
{
	char FILENAME[54];
	int *arr;
	int arr_len; // Nanti diubah via memory address

	printf("Masukkan nama file yang ingin memiliki data untuk di-sort: ");
    scanf("%s", FILENAME);

	arr = read_input(FILENAME, &arr_len);
	printf("Length of the Array: %d\n\n", arr_len);

	printf("Given array is \n");
	printArray(arr, arr_len);

	mergeSort(arr, 0, arr_len - 1);

	printf("\nSorted array is \n");
	printArray(arr, arr_len);

	write_to_txt(arr, arr_len);
}

File: 2 - Sorting/Python/merge_sort.py
import os

# Python program for implementation of MergeSort
def mergeSort(arr):
	if len(arr) > 1:
		# Finding the mid of the array
		mid = len(arr)//2

		# Dividing the array elements
		L = arr[:mid]

		# into 2 halves
		R = arr[mid:]

		# Sorting the first half
		mergeSort(L)

		# Sorting the second half
		mergeSort(R)

		i = j = k = 0

		# Copy data to temp arrays L[] and R[]
		while i < len(L) and j < len(R):
			if L[i] < R[j]:
				arr[k] = L[i]
				i += 1
			else:
				arr[k] = R[j]
				j += 1
			k += 1

		# Checking if any element was left
		while i < len(L):
			arr[k] = L[i]
			i += 1
			k += 1

		while j < len(R):
			arr[k] = R[j]
			j += 1
			k += 1

# Code to print the list
def printList(arr):
	for i in range(len(arr)):
		print(arr[i], end=" ") # syntax python 3.x
	print()

# Driver Code
if __name__ == '__main__':
    # arr = [12, 11, 13, 5, 6, 7]
    
    # INPUT_FILENAME = raw_input("Masukkan nama file yang akan isi datanya akan di-sort (tanpa .txt): ") # python 3.x
    INPUT_FILENAME = input("Masukkan nama file yang akan isi datanya akan di-sort (tanpa .txt): ") # python 3.x
    INPUT_FILENAME += '.txt'
    with open(INPUT_FILENAME, "r") as f:
        row_data = f.read()
        arr = row_data.split(',')
        for i in range(len(arr)):
            try:
                arr[i] = int(arr[i])
            except:
                print("Ada data yang tidak sesuai format")
                print('arr[{0}] = {i}'.format(i, arr[i]))

    print("Given array is", end="\n")
    printList(arr)
    mergeSort(arr)
    print("Sorted array is: ", end="\n")
    printList(arr)

    # WRITE DATA KE FILE BARU
    # FILENAME = raw_input("Masukkan nama file yang akan menampung data yang telah di sort (tanpa .txt): ") # python 2.x
    FILENAME = input("Masukkan nama file yang akan menampung data yang telah di-sort (tanpa .txt): ") # python 3.x
    FILENAME += '.txt'
    data_for_txt = ''
    for i in range(len(arr)):
        data_for_txt += "{0},".format(arr[i])

    PATH = os.getcwd()
    with open(os.path.join(PATH, FILENAME), "w") as f:
        f.write(data_for_txt)

# Most of the code is contributed by Mayank Khanna